pkg/fuzzer/fuzzer.go

```diff
@@ -0,0 +1,298 @@
+	minPrio := zeroPrio
+	if nextExec != nil {
+		minPrio = nextExec.prio
+	}
+	betterJob := fuzzer.genBetterJob(minPrio)
+	if betterJob != nil {
+		fuzzer.Logf(1, "start job of type %T", betterJob)
+		go betterJob.run(fuzzer)
+	}
+	if nextExec != nil {
+		return nextExec.value
+	}
+	// We may end up in a deadlock if the new job is waiting
+	// for something else. But at least so far no job does
+	// something like that.
+	req := fuzzer.nextExec.popWait().value
+	req.sent()
+	return req
+}
+
+func (fuzzer *Fuzzer) genBetterJob(minPrio priority) job {
+	// Candidate job is somewhat special -- there's an external
+	// source of them and yet we don't want to keep the queue too big.
+	// At least until pkg/fuzzer resides on syz-fuzzer.
+	if item := fuzzer.nextJob.pop(
+		minPrio.maxWith(candidateJobPrio)); item != nil {
+		return item.value
+	}
+	if minPrio < candidateJobPrio {
+		select {
+		case candidate := <-fuzzer.Config.Candidates:
+			return newCandidateJob(candidate)
+		default:
+		}
+	}
+	if item := fuzzer.nextJob.pop(minPrio); item != nil {
+		return item.value
+	}
+	if minPrio != zeroPrio {
+		// Don't create jobs to fill out the pipeline.
+		return nil
+	}
+	// Either generate a new input or mutate an existing one.
+	mutateRate := 0.95
+	if !fuzzer.Config.Coverage {
+		// If we don't have real coverage signal, generate programs
+		// more frequently because fallback signal is weak.
+		mutateRate = 0.5
+	}
+	rnd := fuzzer.rand()
+	if rnd.Float64() < mutateRate {
+		if p := fuzzer.Corpus.chooseProgram(rnd); p != nil {
+			return &smashJob{
+				p:     p,
+				short: true,
+			}
+		}
+	}
+	return &genJob{}
+}
+
+func (fuzzer *Fuzzer) Logf(level int, msg string, args ...interface{}) {
+	if fuzzer.Config.Logf == nil {
+		return
+	}
+	fuzzer.Config.Logf(level, msg, args...)
+}
+
+func (fuzzer *Fuzzer) ChoiceTable() *prog.ChoiceTable {
+	fuzzer.ctMu.Lock()
+	defer fuzzer.ctMu.Unlock()
+
+	// TODO: make the step be dependent on the corpus size.
+	const regenerateEveryProgs = 100
+
+	progsCount := len(fuzzer.Corpus.Programs())
+	if fuzzer.ct != nil &&
+		progsCount < fuzzer.ctProgs+regenerateEveryProgs {
+		return fuzzer.ct
+	}
+	fuzzer.ct = fuzzer.target.BuildChoiceTable(fuzzer.Corpus.Programs(),
+		fuzzer.Config.EnabledCalls)
+	fuzzer.ctProgs = progsCount
+	return fuzzer.ct
+}
+
+// Helper methods for jobs.
+
+func (fuzzer *Fuzzer) queueJob(newJob job) {
+	fuzzer.Logf(1, "queued a job of type %T", newJob)
+	fuzzer.nextJob.push(&priorityQueueItem[job]{
+		value: newJob, prio: newJob.priority(),
+	})
+}
+
+func (fuzzer *Fuzzer) rand() *rand.Rand {
+	fuzzer.mu.Lock()
+	seed := fuzzer.rnd.Int63()
+	fuzzer.mu.Unlock()
+	return rand.New(rand.NewSource(seed))
+}
+
+// Later we might split this into exec() and wait() to let jobs
+// post async exec requests.
+func (fuzzer *Fuzzer) execWait(job job, req *Request) *Result {
+	if req.stat == "" {
+		panic("Request.Stat field must be set")
+	}
+	req.resultCond = sync.NewCond(&sync.Mutex{})
```

Comment:
This is overly complex and generates lots of garbage.

Condition variables in Go are a kind of a red flag for me, they are harder to use.
Instead of the condition variable + goroutine that resends to the channel, Done should just send result to a channel.

Timeout is used only for bug detection. Let's think how to remove it.
When we move fuzzer into syz-manager, we will need a map of pending requests (to be able to map integer request ID to the object + to handle VM crashes, see how syz-verifier does it). So we can add such map now and remember time when the request was started. And then have a goroutine that runs every 10 mins or so and checks timed out requests.

Reply:
Changed the logic, now execution leaks are calculated in a separate goroutine.